File: lib/domain/cultivation_plan/interactors/task_schedule_timeline_interactor.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Interactors
      class TaskScheduleTimelineInteractor
        def initialize(
          output_port:,
          user_id:,
          plan_id:,
          timeline_read_gateway:,
          cultivation_plan_gateway:,
          translator:,
          logger:,
          user_lookup:,
          clock:
        )
          raise ArgumentError, "clock must respond to :today" unless clock.respond_to?(:today)

          @output_port = output_port
          @user_id = user_id
          @plan_id = plan_id
          @timeline_read_gateway = timeline_read_gateway
          @cultivation_plan_gateway = cultivation_plan_gateway
          @translator = translator
          @logger = logger
          @user_lookup = user_lookup
          @clock = clock
        end

        def call
          user = begin
            @user_lookup.find(@user_id)
          rescue Domain::Shared::Exceptions::RecordNotFound
            @logger.warn("[TaskScheduleTimelineInteractor] user_record_not_found user_id=#{@user_id.inspect}")
            @output_port.on_failure(Domain::Shared::Dtos::Error.new(@translator.t("plans.errors.session_invalid")))
            return
          end

          unless TaskSchedulePrivatePlanAccess.access_allowed?(
            plan_gateway: @cultivation_plan_gateway, plan_id: @plan_id, user_id: user.id
          )
            raise Domain::Shared::Exceptions::RecordNotFound, "Cultivation plan not found"
          end

          read_model = Mappers::TaskScheduleTimelineReadSnapshotMapper.load_snapshot(
            read_gateway: @timeline_read_gateway,
            plan_id: @plan_id
          )
          dto = Mappers::TaskScheduleTimelineMapper.call(read_model, today: @clock.today)
          @output_port.on_success(dto)
        rescue NoMethodError, NameError, ArgumentError, SyntaxError
          raise
        rescue Domain::Shared::Exceptions::PersistenceFailed => e
          log_interactor_error(e)
          raise
        rescue Domain::Shared::Exceptions::RecordNotFound => e
          @logger.warn("[TaskScheduleTimelineInteractor] record_not_found: #{e.class}: #{e.message}")
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(@translator.t("plans.errors.not_found")))
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @logger.warn("[TaskScheduleTimelineInteractor] record_invalid: #{e.class}: #{e.message}")
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end

        private

        def log_interactor_error(error)
          bt = error.backtrace&.first(20)&.join("\n").to_s
          @logger.error(
            "[TaskScheduleTimelineInteractor] #{error.class}: #{error.message}\n/backtrace:\n#{bt}"
          )
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/interactors/task_schedule_timeline_interactor.rs
//! Ruby: `Domain::CultivationPlan::Interactors::TaskScheduleTimelineInteractor`

use crate::cultivation_plan::gateways::{CultivationPlanGateway, CultivationPlanPrivateReadGateway};
use crate::cultivation_plan::mappers::task_schedule_timeline_mapper::TaskScheduleTimelineMapper;
use crate::cultivation_plan::ports::TaskScheduleTimelineOutputPort;
use crate::cultivation_plan::interactors::task_schedule_private_plan_access;
use crate::shared::dtos::Error;
use crate::shared::exceptions::{RecordInvalidError, RecordNotFoundError};
use crate::shared::gateways::UserLookupGateway;
use crate::shared::ports::{ClockPort, LoggerPort, TranslatorPort};

pub struct TaskScheduleTimelineInteractor<'a, O, G, R, U, L, T, C> {
    output_port: &'a mut O,
    user_id: i64,
    plan_id: i64,
    private_read_gateway: &'a R,
    cultivation_plan_gateway: &'a G,
    translator: &'a T,
    logger: &'a L,
    user_lookup: &'a U,
    clock: &'a C,
}

impl<'a, O, G, R, U, L, T, C> TaskScheduleTimelineInteractor<'a, O, G, R, U, L, T, C>
where
    O: TaskScheduleTimelineOutputPort,
    G: CultivationPlanGateway,
    R: CultivationPlanPrivateReadGateway,
    U: UserLookupGateway,
    L: LoggerPort,
    T: TranslatorPort,
    C: ClockPort,
{
    pub fn new(
        output_port: &'a mut O,
        user_id: i64,
        plan_id: i64,
        private_read_gateway: &'a R,
        cultivation_plan_gateway: &'a G,
        translator: &'a T,
        logger: &'a L,
        user_lookup: &'a U,
        clock: &'a C,
    ) -> Self {
        Self {
            output_port,
            user_id,
            plan_id,
            private_read_gateway,
            cultivation_plan_gateway,
            translator,
            logger,
            user_lookup,
            clock,
        }
    }

    pub fn call(&mut self) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_lookup.find(self.user_id);

        if !task_schedule_private_plan_access::access_allowed(
            self.cultivation_plan_gateway,
            self.plan_id,
            user.id,
        ) {
            return Err(Box::new(RecordNotFoundError));
        }

        match self
            .private_read_gateway
            .find_task_schedule_timeline_by_plan_id(self.plan_id)
        {
            Ok(read_model) => {
                let dto = TaskScheduleTimelineMapper::call(read_model, self.clock.today());
                self.output_port.on_success(dto);
                Ok(())
            }
            Err(err) if err.downcast_ref::<RecordNotFoundError>().is_some() => {
                self.logger.warn("[TaskScheduleTimelineInteractor] record_not_found");
                self.output_port
                    .on_failure(Error::new(self.translator.t("plans.errors.not_found", &Default::default())));
                Ok(())
            }
            Err(err) if err.downcast_ref::<RecordInvalidError>().is_some() => {
                let invalid = err.downcast_ref::<RecordInvalidError>().unwrap();
                self.output_port.on_failure(Error::new(invalid.to_string()));
                Ok(())
            }
            Err(err) => Err(err),
        }
    }
}
